crates/wasmtime/src/module/registry.rs

```diff
@@ -48,66 +43,91 @@ impl StoreFrameInfo {
         self.module(pc)?.lookup_trap_info(pc)
     }
 
-    fn module(&self, pc: usize) -> Option<&ModuleFrameInfo> {
-        let (end, info) = self.ranges.range(pc..).next()?;
+    /// Looks up a stack map from a program counter.
+    pub fn lookup_stack_map<'a>(&'a self, pc: usize) -> Option<&'a ir::StackMap> {
+        self.module(pc)?.lookup_stack_map(pc)
+    }
+
+    fn module(&self, pc: usize) -> Option<&RegisteredModule> {
+        let (end, info) = self.0.range(pc..).next()?;
         if pc < info.start || *end < pc {
             return None;
         }
 
         Some(info)
     }
 
-    /// Registers a new compiled module's frame information.
-    pub fn register(&mut self, module: &Arc<CompiledModule>) {
-        let (start, end) = module.code().range();
+    /// Registers a new module with the registry.
+    pub fn register(&mut self, module: &Module) {
+        let compiled_module = module.compiled_module();
+        let (start, end) = compiled_module.code().range();
 
-        // Ignore modules with no code or finished functions
-        if start == end || module.finished_functions().is_empty() {
+        // Ignore modules with no code, finished functions, or if the module is already registered
+        if start == end || compiled_module.finished_functions().is_empty() {
             return;
         }
 
         // The module code range is exclusive for end, so make it inclusive as it
         // may be a valid PC value
         let end = end - 1;
 
+        if self.0.get(&end).is_some() {
+            return;
+        }
+
         // Assert that this module's code doesn't collide with any other registered modules
-        if let Some((_, prev)) = self.ranges.range(end..).next() {
+        if let Some((_, prev)) = self.0.range(end..).next() {
             assert!(prev.start > end);
         }
-        if let Some((prev_end, _)) = self.ranges.range(..=start).next_back() {
+
+        if let Some((prev_end, _)) = self.0.range(..=start).next_back() {
             assert!(*prev_end < start);
         }
 
-        let prev = self.ranges.insert(
+        let prev = self.0.insert(
             end,
-            ModuleFrameInfo {
+            RegisteredModule {
                 start,
-                module: module.clone(),
+                module: compiled_module.clone(),
+                signatures: module.shared_signatures().clone(),
             },
         );
         assert!(prev.is_none());
 
-        GLOBAL_INFO.lock().unwrap().register(start, end, module);
+        GLOBAL_MODULES.lock().unwrap().register(start, end, module);
+    }
+
+    /// Looks up a trampoline from a shared signature index.
+    ///
+    /// This will search all modules associated with the store for a suitable trampoline
+    /// given the shared signature index.
+    pub fn lookup_trampoline(&self, index: VMSharedSignatureIndex) -> Option<VMTrampoline> {
+        for (_, m) in &self.0 {
```

Comment:
FWIW this seems like it could be prohibitively slow or a source of accidental n^2 things. I'm basically worried that this linear search could take a surprisingly long time in situations where a store has a number of modules or something like that (especially with module linking and whatnot).

I think this is only called from `Func::from_wasmtime_function` so the impact is at least somewhat isolated, but even then that's a pretty core piece of functionality that I'm not entirely sure how would be refactored.

Do you have thoughts on how this might be possible to remove the loop?

Reply:
If the shared signature space were guaranteed contiguous for a module, we could easily keep a range map around for at least O(log(n)) lookup, but the free list maintained by the signature registry precludes that.

Let me think on this more.

Reply:
Since this is only called from `from_wasmtime_function` could we perhaps use the entry point of the wasm function provided there for lookup? If this took a `VMCallerCheckedAnyfunc` instead of a `VMSharedSignatureIndex` we could do a pc lookup to find which module has that pc, then we look up in that module's trampoline map directly?

Reply:
OooOoo excellent idea :) I'll make that change.